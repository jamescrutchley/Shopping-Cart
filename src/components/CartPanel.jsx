import { useContext, useEffect } from "react";
import styles from "../styles/SearchPanel.module.css";
import CartContext from "../context/CartContext";
import CartCard from "./CartCard";
import sumPrice from "../helpers/sumPrice";
import { DollarCents } from "../helpers/DollarCents";
import { useClickAway } from "react-use";

import buttonStyles from "../styles/Buttons.module.css";
import { useState, useRef } from "react";

const CartPanel = ({ clickBack }) => {
  const [fadeOutClass, setFadeOutClass] = useState(null);
  const { cartArray, addToCart } = useContext(CartContext);
  const shippingPrice =
    sumPrice(...cartArray.map((item) => 1 * item.quantity)) / 2;
  const ref = useRef(null);

  useClickAway(ref, () => {
    notifyClickBack();
  });

  const notifyClickBack = () => {
    setFadeOutClass(true);
    setTimeout(() => {
      clickBack(null); // Call clickBack after 0.2s
    }, 150);
  };

  return (
    <section
      className={`${styles.panelWrapper} ${fadeOutClass ? styles.fade : ""}`}
    >
      <div
        ref={ref}
        className={`${styles.cart} ${styles.searchContainer} ${
          fadeOutClass ? styles.fade : ""
        }`}
      >
        <button onClick={() => notifyClickBack()} className={styles.back}>
          Back
        </button>
        <div className={styles.cartHeaders}>
          <p>Item</p>
          <p className={styles.price}>Price</p>
          <p className={styles.quantity}>Quantity</p>
          <p>Total</p>
        </div>
        <h2>Your Cart</h2>
        <div className={styles.cartWrapper}>
          <section className={styles.cartContainer}>
            {cartArray.map((data, index) => (
              <CartCard key={index} data={data} />
            ))}
          </section>
          <section className={styles.orderSummaryContainer}>
            <div>
              <p>Subtotal</p>
              <p>
                {DollarCents(
                  sumPrice(
                    ...cartArray.map((item) => item.price * item.quantity),
                  ),
                )}
              </p>
            </div>
            <div>
              <p>Shipping</p>
              <p>{DollarCents(shippingPrice)}</p>
            </div>
            <div>
              <p>Total</p>
              <p>
                {DollarCents(
                  sumPrice(
                    ...cartArray.map((item) => item.price * item.quantity),
                  ) + shippingPrice,
                )}
              </p>
            </div>
          </section>
        </div>
      </div>
    </section>
  );
};

export default CartPanel;
