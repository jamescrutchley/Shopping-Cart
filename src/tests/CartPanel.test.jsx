import CartContext from '../context/CartContext'; 
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import CartPanel from '../components/CartPanel';
import { renderWithCartContext, mockCartContext } from './mockContext/mockContext';


import { describe, it, expect, vitest, vi, beforeEach } from "vitest";
import { BrowserRouter } from "react-router-dom";
import { fakeStore } from "../mocks/copiedObj.js";


  const callBack = vi.fn();

describe('CartPanel behaves correctly', () => {

    it('panel renders', () => {
        const { asFragment } = renderWithCartContext(
            <CartPanel clickBack={callBack} />,
            mockCartContext
          );
          expect(asFragment()).toMatchSnapshot();
      
    })
    it('renders right number of cartcard items', () => {
        renderWithCartContext(
            <CartPanel clickBack={callBack} />,
            mockCartContext
          );

          const cartItems = screen.getAllByTestId('card');
      
          expect(cartItems).toHaveLength(4);
    })

    //trigger useClickAway hook?

    // it('clickaway fires', async () => {
    //     renderWithCartContext(
    //         <CartPanel clickBack={callBack} />,
    //         mockCartContext
    //       );
    //     const wrapper = screen.getByTestId('panel-wrapper');
    //     expect(wrapper).toBeInTheDocument();
    //     await waitFor(() => {
    //         expect(callBack).toHaveBeenCalled();
    //     })
    // })
})