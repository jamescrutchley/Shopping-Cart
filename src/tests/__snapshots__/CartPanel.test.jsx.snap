// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CartPanel behaves correctly > panel renders 1`] = `
<DocumentFragment>
  <div
    class="_panelWrapper_7f3fc5 "
    data-testid="panel-wrapper"
  >
    <div
      class="_cart_7f3fc5 _searchContainer_7f3fc5 "
    >
      <button
        class="_back_7f3fc5"
      >
        Back
      </button>
      <div
        class="_cartHeaders_7f3fc5"
      >
        <p>
          Item
        </p>
        <p
          class="_price_7f3fc5"
        >
          Price
        </p>
        <p
          class="_quantity_7f3fc5"
        >
          Quantity
        </p>
        <p>
          Total
        </p>
      </div>
      <h2>
        Your Cart
      </h2>
      <div
        class="_cartWrapper_7f3fc5"
      >
        <section
          class="_cartContainer_7f3fc5"
        >
          <article
            class="_card_d72dd6"
            data-testid="card"
          >
            <div
              class="_cardImg_d72dd6"
            >
              <img
                alt="product thumbnail"
                class="_thumbnail_d72dd6"
                src="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"
              />
            </div>
            <div
              class="_cardInfo_d72dd6"
            >
              <div
                class="_titleContainer_d72dd6"
              >
                <p>
                  Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
                </p>
              </div>
              <div
                class="_priceContainer_d72dd6"
              >
                <p>
                  $109.95
                </p>
              </div>
              <div
                class="_quantityControls_4bc385"
              >
                <label
                  for="quantity"
                >
                  Qty:
                </label>
                <div
                  class="_controlsWrapper_4bc385"
                >
                  <div
                    class="_decrementContainer_4bc385"
                  >
                    <button>
                      -
                    </button>
                  </div>
                  <div
                    class="_clickableInputContainer_4bc385"
                  >
                    <input
                      class="_clickableInput_4bc385"
                      id="quantity"
                      max="9"
                      min="1"
                      name="quantity"
                      type="quantity"
                      value="4"
                    />
                  </div>
                  <div
                    class="_incrementContainer_4bc385"
                  >
                    <button>
                      +
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="_priceTotalContainer_d72dd6"
              >
                <p>
                  $439.80
                </p>
              </div>
              <div
                class="_productControlsContainer_d72dd6"
              />
            </div>
          </article>
          <article
            class="_card_d72dd6"
            data-testid="card"
          >
            <div
              class="_cardImg_d72dd6"
            >
              <img
                alt="product thumbnail"
                class="_thumbnail_d72dd6"
                src="https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg"
              />
            </div>
            <div
              class="_cardInfo_d72dd6"
            >
              <div
                class="_titleContainer_d72dd6"
              >
                <p>
                  Mens Casual Premium Slim Fit T-Shirts 
                </p>
              </div>
              <div
                class="_priceContainer_d72dd6"
              >
                <p>
                  $22.30
                </p>
              </div>
              <div
                class="_quantityControls_4bc385"
              >
                <label
                  for="quantity"
                >
                  Qty:
                </label>
                <div
                  class="_controlsWrapper_4bc385"
                >
                  <div
                    class="_decrementContainer_4bc385"
                  >
                    <button>
                      -
                    </button>
                  </div>
                  <div
                    class="_clickableInputContainer_4bc385"
                  >
                    <input
                      class="_clickableInput_4bc385"
                      id="quantity"
                      max="9"
                      min="1"
                      name="quantity"
                      type="quantity"
                      value="4"
                    />
                  </div>
                  <div
                    class="_incrementContainer_4bc385"
                  >
                    <button>
                      +
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="_priceTotalContainer_d72dd6"
              >
                <p>
                  $89.20
                </p>
              </div>
              <div
                class="_productControlsContainer_d72dd6"
              />
            </div>
          </article>
          <article
            class="_card_d72dd6"
            data-testid="card"
          >
            <div
              class="_cardImg_d72dd6"
            >
              <img
                alt="product thumbnail"
                class="_thumbnail_d72dd6"
                src="https://fakestoreapi.com/img/71li-ujtlUL._AC_UX679_.jpg"
              />
            </div>
            <div
              class="_cardInfo_d72dd6"
            >
              <div
                class="_titleContainer_d72dd6"
              >
                <p>
                  Mens Cotton Jacket
                </p>
              </div>
              <div
                class="_priceContainer_d72dd6"
              >
                <p>
                  $55.99
                </p>
              </div>
              <div
                class="_quantityControls_4bc385"
              >
                <label
                  for="quantity"
                >
                  Qty:
                </label>
                <div
                  class="_controlsWrapper_4bc385"
                >
                  <div
                    class="_decrementContainer_4bc385"
                  >
                    <button>
                      -
                    </button>
                  </div>
                  <div
                    class="_clickableInputContainer_4bc385"
                  >
                    <input
                      class="_clickableInput_4bc385"
                      id="quantity"
                      max="9"
                      min="1"
                      name="quantity"
                      type="quantity"
                      value="2"
                    />
                  </div>
                  <div
                    class="_incrementContainer_4bc385"
                  >
                    <button>
                      +
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="_priceTotalContainer_d72dd6"
              >
                <p>
                  $111.98
                </p>
              </div>
              <div
                class="_productControlsContainer_d72dd6"
              />
            </div>
          </article>
          <article
            class="_card_d72dd6"
            data-testid="card"
          >
            <div
              class="_cardImg_d72dd6"
            >
              <img
                alt="product thumbnail"
                class="_thumbnail_d72dd6"
                src="https://fakestoreapi.com/img/71YXzeOuslL._AC_UY879_.jpg"
              />
            </div>
            <div
              class="_cardInfo_d72dd6"
            >
              <div
                class="_titleContainer_d72dd6"
              >
                <p>
                  Mens Casual Slim Fit
                </p>
              </div>
              <div
                class="_priceContainer_d72dd6"
              >
                <p>
                  $15.99
                </p>
              </div>
              <div
                class="_quantityControls_4bc385"
              >
                <label
                  for="quantity"
                >
                  Qty:
                </label>
                <div
                  class="_controlsWrapper_4bc385"
                >
                  <div
                    class="_decrementContainer_4bc385"
                  >
                    <button>
                      -
                    </button>
                  </div>
                  <div
                    class="_clickableInputContainer_4bc385"
                  >
                    <input
                      class="_clickableInput_4bc385"
                      id="quantity"
                      max="9"
                      min="1"
                      name="quantity"
                      type="quantity"
                      value="1"
                    />
                  </div>
                  <div
                    class="_incrementContainer_4bc385"
                  >
                    <button>
                      +
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="_priceTotalContainer_d72dd6"
              >
                <p>
                  $15.99
                </p>
              </div>
              <div
                class="_productControlsContainer_d72dd6"
              />
            </div>
          </article>
        </section>
        <section
          class="_orderSummaryContainer_7f3fc5"
        >
          <div>
            <p>
              Subtotal
            </p>
            <p>
              $656.97
            </p>
          </div>
          <div>
            <p>
              Shipping
            </p>
            <p>
              $5.50
            </p>
          </div>
          <div>
            <p>
              Total
            </p>
            <p>
              $662.47
            </p>
          </div>
        </section>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
